Latest_stable_version

ConstVol opened the pull request:
- sorting csv IMDB data by movie genre
- inline keyboard to choose movie genre
- rutracker API integration
  gem 'rutracker_api' out of date - need to change login link.
- cool oneliner functions

Probably oneliners are difficult to read but i wanted to try it.

ConstVol committed:
- inline keyboard to choose movie genre
- IMDB link instead of movie title
- rutracker API integration
- cool oneliner functions

anatoliliotych commented on bot.rb:
```diff
@@ -0,0 +1,46 @@
+class IMDB_Helper
+
+  size_to_kb = {'KB' => 1, 'MB' => 1024, 'GB' => 1024*1024} # convert everything to KB in order compare sizes
+  client = RutrackerApi.new('username', 'password')
```
where do you store username and password?

ConstVol replied:
I'll create constants on top

anatoliliotych commented on bot.rb:
```diff
@@ -0,0 +1,46 @@
+  buttons = []
+  genres.keys.each { |genre| buttons << Telegram::Bot::Types::InlineKeyboardButton.new(text: genre, callback_data: genre)}
+
+  Telegram::Bot::Client.run('telegram_bot_token') do |bot|
```
what is the way to set token?

ConstVol replied:
I'll create constant on top

anatoliliotych replied:
use .env file for this with gem dotenv

anatoliliotych commented on bot.rb:
```diff
@@ -0,0 +1,46 @@
+require 'csv'
+require 'rutracker_api'
+
+class IMDB_Helper
```
how do you use this class? how to launch bot?

ConstVol replied:
I don't use it. I'll get rid of it.

anatoliliotych commented on bot.rb:
```diff
@@ -0,0 +1,46 @@
+  csv = CSV.new(File.read('./data/movie_metadata.csv'), :headers => true)
+
+  # sorting IMDB data in {'genre' => [{'title' => 'imbd_link'}, ...], ...} form. In one line just for fun;)
+  csv.each {|row| row.to_hash['genres'].split('|').each {|genre| genres[genre].push({row.to_hash['movie_title'] => row.to_hash['movie_imdb_link']})} if row.to_hash['imdb_score'].to_f > 7}
```
too long string! try to make code more readable

ConstVol replied:
will revert to a separete method

ConstVol committed:
Progect structure refactoring

- Deleted class IMDB_Helper
- added constants for username, password and token
- added method sort_by_genres

IMDB_Helper was useless.
Constants needed to store username, password and token.
Method sort_by_genres was created to make code more readable

anatoliliotych commented on bot.rb:
```diff
@@ -0,0 +1,56 @@
+require 'csv'
+require 'rutracker_api'
+
+USERNAME = ''
```
will this code work? or you will need to put creds here before launch? :)

use .env file with dotenv gem

anatoliliotych commented on bot.rb:
```diff
@@ -0,0 +1,56 @@
+# sorting IMDB data in {'genre' => [{'title' => 'imbd_link'}, ...], ...} form
+  csv.each do |row|
+    movie = row.to_hash
+    movie['genres'].split('|').each do|genre|
```
style guide!

anatoliliotych commented on bot.rb:
```diff
@@ -0,0 +1,56 @@
+  csv.each do |row|
+    movie = row.to_hash
+    movie['genres'].split('|').each do|genre|
+      if movie['imdb_score'].to_f > 7
```
to float and comparing with integer

anatoliliotych commented on bot.rb:
```diff
@@ -0,0 +1,56 @@
+        search_result = client.search(term:movie.keys[0])
+
+        #searching on rutracker by movie title(checking if response isn't {:error => 'Not found'}) -> sorting response by size -> getting the last one(biggest one)
+        torrent_id = client.search(term:movie.keys[0]).sort_by{|torrent| torrent[:size].split(" ")[0].to_f * size_to_kb[torrent[:size].split(" ")[1]]}.last[:torrent_id] unless search_result.is_a? Hash
```
can you read this? try to make code readable

anatoliliotych commented on bot.rb:
```diff
@@ -0,0 +1,56 @@
+genres.keys.each { |genre| buttons << Telegram::Bot::Types::InlineKeyboardButton.new(text: genre, callback_data: genre)}
+
+Telegram::Bot::Client.run(TOKEN) do |bot|
+
```
remove empty lines 
read style guide about this

ConstVol committed:
Refactoring according to pull request comments

- used dotenv gem to hide private info
- add RutrackeHandler and Sorter classes
- add modified RurtrackerApi gem

anatoliliotych commented on bot.rb:
```diff
@@ -0,0 +1,24 @@
+  bot.listen do |message|
+    case
+      when message.is_a?(Telegram::Bot::Types::CallbackQuery)
+        movie = genres[message.data].sample # get random movie by genre in ['title' => 'imdb_link'] form
```
If comments are obvious, I'd skip them.

ConstVol replied:
For some bodies it may not be so obvious

anatoliliotych replied:
try to write obvious and readable code :)

anatoliliotych commented on rutracker_api.rb:
```diff
@@ -0,0 +1,78 @@
+
+  def find_user(nick)
+    # TODO: add parser for user pages
+    @agent.get 'http://rutracker.org/forum/profile.php?mode=viewprofile&u=' + nick
```
URL to constant! you have lots of common in urls, try to prepare method which removes duplication

anatoliliotych commented on rutracker_api.rb:
```diff
@@ -0,0 +1,78 @@
+  def login(username, pass)
+    @agent.post(LOGIN_PAGE, login_username: username,
+                login_password: pass, login: 'Вход')
+
```
remove empty line

anatoliliotych commented on rutracker_api.rb:
```diff
@@ -0,0 +1,78 @@
+
+  def parse_search
+    @agent.page.search("//table[@class='forumline tablesorter']/tbody/tr").map do |row|
+      if row.css(".row1").text =~ /Не найдено/
```
make oneliner

ConstVol replied:
It would be disrespectful to the creator of rutracker_api gem:)

anatoliliotych replied:
please make : 
```
return {} if condition
```

anatoliliotych commented on rutracker_handler.rb:
```diff
@@ -0,0 +1,21 @@
+    unless search_result.is_a? Hash
+      sorted_by_size = sorter.sort_by_torrent_size(search_result)
+      torrent_id = sorted_by_size.last[:torrent_id]
+      "https://rutracker.org/forum/viewtopic.php?t=#{torrent_id}"
```
url to constant

anatoliliotych commented on sorter.rb:
```diff
@@ -0,0 +1,33 @@
+class Sorter
+
+  def initialize
+    @file_name = './data/movie_metadata.csv'
```
filename should be in constant

anatoliliotych commented on sorter.rb:
```diff
@@ -0,0 +1,33 @@
+
+  def initialize
+    @file_name = './data/movie_metadata.csv'
+    @size_to_kb = {'KB' => 1, 'MB' => 1024, 'GB' => 1024*1024} # convert everything to KB in order compare sizes
```
use constant

anatoliliotych replied:
check style

anatoliliotych commented on rutracker_api.rb:
```diff
@@ -0,0 +1,78 @@
+  ORDER_OPTIONS = { date: 1,  name: 2, downloads: 4, shows: 6, seeders: 10,
+                    leechers: 11, size: 7, last_post: 8, speed_up: 12,
+                    speed_down: 13, message_count: 5, last_seed: 9 }
+  SORT_OPTIONS = {asc: 1, desc: 2}
```
style guide!

anatoliliotych commented on rutracker_api.rb:
```diff
@@ -0,0 +1,78 @@
+  def parse_search
+    @agent.page.search("//table[@class='forumline tablesorter']/tbody/tr").map do |row|
+      if row.css(".row1").text =~ /Не найдено/
+        return {error: 'Not found'}
```
style guide

anatoliliotych reviewed (commented):
respond to code review!

ConstVol committed:
LGTM

anatoliliotych commented on bot.rb:
```diff
@@ -0,0 +1,24 @@
+require 'telegram/bot'
+require 'dotenv/load'
```
add .env.example file

anatoliliotych commented on rutracker_api.rb:
```diff
@@ -0,0 +1,78 @@
+    prepared = { f: options[:category], nm: options[:term],
+                 s: SORT_OPTIONS[options[:sort]], o: ORDER_OPTIONS[options[:order_by]] }
+
+    query = SEARCH_PAGE + '?'
```
is it worth to have this + ?

anatoliliotych reviewed (changes requested):
Fix commit messages

ConstVol committed:
added .env.example file, some rutracker_api refactoring. this is not a challenge:)

anatoliliotych reviewed (approved):
LGTM
